specification/dartLangSpec.tex

```diff
@@ -2379,28 +2382,40 @@ \subsubsection{The Method \code{noSuchMethod}}
 one of the following is true:
 
 \begin{itemize}
-\item $C$ has a non-trivial \code{noSuchMethod},
-  the interface of $C$ contains $m$,
-  and $C$ has no concrete declaration of $m$
-  (\commentary{that is, no member $m$ is declared or inherited by $C$}).
+\item \textbf{Requested in program:}
+  $C$ has a non-trivial \code{noSuchMethod},
+  the interface of $C$ contains a member signature $S$ named $m$,
+  and $C$ has no concrete member named $m$ that correctly overrides $S$
+  (\commentary{%
+    that is, no member named $m$ is declared or inherited by $C$,
+    or one is inherited, but it does not have the required signature%
+  }).
 \item
   % Inaccessible private methods are not present in the interface of a class,
   % so we need to find a class that can access $m$.
+  \textbf{Forced by privacy:}
   There exists a direct or indirect superinterface
   $D$ of $C$ which is declared in the library $L_2$,
   the interface of $D$ contains $m$
   (\commentary{which implies that $m$ is accessible to $L_2$}),
   $m$ is inaccessible to $L$,
   and no superclass of $C$ has
-  a concrete declaration of $m$ accessible to $L_2$.
+  a concrete member of $m$ accessible to $L_2$.
```

Comment:
... 
This doesn't work with "member of $m$" instead of "declaration of $m$".

I would say:
> No superclass of $C$ declares a concrete member implementing $m$,

It's still not perfect.


More fundamentally, I guess: What is a "member"?

As I understand it:

A class *declaration* contains member *declarations*, and these can be concrete or abstract. A class declaration induces (at least) a class and an interface (modulo generics as usual).

A class contains members induced by the declarations in the class declaration or inherited from the superclass, or introduced implicitly in other ways like NSM-forwarding (or coming from mixins in some cases, or ... etc).

An interface contains member signatures. Each signature with a private name is also associated with a library and it is only "accessible" inside that same library. There can be more than one signature with the same name, but only if they are private to different libraries.

Class members can implement interface member signatures with the same accessible name, or fail to do so if their function signature is not sufficient.

So, a NSM forwarder added to a class is a class *member* (concrete, but that's a pleonasm), and here we say that such a forwarder would prevent a NSM forwarder in a subclass.
```dart
/// library L2
class C {
   noSuchMethod(i) {}
   // Gets noSuchMethod forwarder member for `foo` interface signature.
   int _foo();
}
abstract class D extends C {
  int _foo([int x]);
}
/// Library L1
class E extends D {
  // Should get noSuchMethod-forwarder for D._foo.
}
```

So, this item does not capture that case unless it does not consider a NSM forwarder as a concrete member (which would probably be an inconsistent point of view). (Or unless it's rewritten as above, so only *declared* concrete members matter).

Reply:
Right, I agree that a declaration is a syntactic entity. It is associated with a static entity which is semantic in the sense that it may have a signature or a type or other properties that are not purely syntactic (that is, not context free, we must do some analysis before we have this data). It is also associated with a dynamic entity (such as a class at run time, which allows us to talk about the value of a type variable at run time, or the value of a static variable, etc).

I also agree that a class interface contains member signatures, and they are specified explicitly already.

I have now defined 'NSM-forwarded in a class $C$ declared in a library $L$' to be a property of a name $m$.

The example you specify was actually omitted, and I have already changed the text accordingly. So we agree about that, too. ;-)

Reply:
Agreed offline.